### PR TITLE
use nbserverproxy for hpc install setup guide

### PR DESCRIPTION
**Pull request by jhamman**
These changes reflect the fact that we don't need to forward a port for the dask distributed dashboard from HPC machines if we install `nbserverproxy`. 

This seems to be working for me but I'd appreciate a few others try it out before we merge this.

@mrocklin - thoughts?

### CODE REVIEW

**Comment by rabernat** [comment]
Maybe @sridge could try this out, since he is pretty active on Cheyenne these days.

File: docs/setup_guides/hpc.rst [hunk 1]
```diff
@@ -117,6 +120,14 @@ securing Jupyter, check out
 `Securing a notebook server <http://jupyter-notebook.readthedocs.io/en/stable/public_server.html#securing-a-notebook-server>`__
 in the Jupyter documentation.
 
+Finally, we may want to configure dask's dashboard to forward through jupyterhub.
```
Comment (mrocklin) [inline docs/setup_guides/hpc.rst]: JupyterHub -> Jupyter

File: docs/setup_guides/hpc.rst [hunk 2]
```diff
@@ -117,6 +120,14 @@ securing Jupyter, check out
...
+This can be done by editing the dask distributed config file, e.g.:
+``.config/dask/distributed.yaml``. In this file, set:
+
+.. code:: python
+
+    diagnostics-link: "../proxy/{port}/status"
```
Comment (mrocklin) [inline docs/setup_guides/hpc.rst]: Classic Jupyter notebooks and JupyterLab have different relative paths.  I wonder if a static path like `/proxy/{port}/status` can be made to work.

File: docs/setup_guides/hpc.rst [hunk 3]
```diff
@@ -41,7 +41,9 @@ Create a new conda environment for our pangeo work:
 ::
 
     conda create -n pangeo -c conda-forge \
-        python=3.6 dask distributed xarray jupyterlab mpi4py dask-jobqueue
+        python=3.6 xarray \
+        jupyterlab nbserverproxy \
+        dask distributed mpi4py dask-jobqueue
```
Comment (mrocklin) [inline docs/setup_guides/hpc.rst]: I take it that we're still using the `dask-mpi` solution here?

Comment (jhamman) [reply docs/setup_guides/hpc.rst]: It is mentioned as an option at the bottom of the page.

**Comment by sridge** [comment]
Just tried with Dask 17.5. Getting a connection refused error in the browser:

```[Errno 111] Connection refused```

 and this in the notebook log:

```[E 09:23:04.954 NotebookApp] 500 GET /proxy/8787/status (10.148.0.5) 2.47ms referer=http://localhost:8888/notebooks/LENS_budget_carbon_example.ipynb```

**Comment by jhamman** [comment]
@edougherty32 - mind taking a look at the changes I just made in 68f7acf. I think they should cover most of questions you ran into in #315.

**Comment by jhamman** [comment]
@sridge -thanks for giving this a try. I may ask you to go through things one more time but I'll need to test things a bit further first.

**Comment by guillaumeeb** [comment]
@jhamman I missed your work here when I opened https://github.com/dask/dask-jobqueue/issues/111. 

I will be happy to give this a try once you feel this is ready!

File: docs/setup_guides/hpc.rst [hunk 4]
```diff
@@ -117,12 +125,20 @@ securing Jupyter, check out
 `Securing a notebook server <http://jupyter-notebook.readthedocs.io/en/stable/public_server.html#securing-a-notebook-server>`__
 in the Jupyter documentation.
 
+Finally, we may want to configure dask's dashboard to forward through Jupyter.
+This can be done by editing the dask distributed config file, e.g.:
+``.config/dask/distributed.yaml``. In this file, set:
+
+.. code:: python
+
+    diagnostics-link: "/proxy/{port}/status"
```
Comment (guillaumeeb) [inline docs/setup_guides/hpc.rst]: Was not sure where to add this, I added it beyond distributed at first, but it was needed at the root of the yaml file.

Comment (guillaumeeb) [reply docs/setup_guides/hpc.rst]: We should also maybe add a full link to where the dashboard is expected to be find, e.g. https://jupyter-server/user/username/proxy/8787/status if I understand correctly.

Comment (jhamman) [reply docs/setup_guides/hpc.rst]: I actually think this still needs to be a relative link, something like `../proxy/{port}/status`. @mrocklin had suggested this absolute path but I don't think this is working for folks in practice. The differences in the url from the standard jupyter notebook interface and jupyter lab interface further complicates this.

**Comment by mrocklin** [comment]
I think that absolute paths are nicer due to the lab/notebook difference
that you mention.  This often means that you need some sort of
specialization though.

File: docs/setup_guides/hpc.rst (hunk 4 above)
Comment (guillaumeeb) [reply docs/setup_guides/hpc.rst]: Neither `../proxy` nor `/proxy` works for me with jupyterhub or manually launched notebook, the same behavior can be observed:
URL for `/proxy`:
````
https://jupyterhub.sis.cnes.fr/user/eynardbg/files/proxy/38238/status
````

URL for `../proxy`:
````
http://node136.sis.cnes.fr:8877/files/git/code-snippets/proxy/46876/status
````

So it always redirect to `notebook/files/`, and then absolute or relative path from where the notebook is located...

Comment (guillaumeeb) [reply docs/setup_guides/hpc.rst]: On standard jupyter notebook, solutions that work for me either from manually launched notebook or Jupyterhub (see https://github.com/dask/dask-jobqueue/pull/144/files) are:
````
distributed.dashboard.link: "/proxy/{port}/status" # for user launched notebook
distributed.dashboard.link: "/user/{JUPYTERHUB_USER}/proxy/{port}/status" # for jupyterhub launched notebook
````

**Comment by mnlevy1981** [comment]
I just want to chime in to say that I don't like

```
cluster = PBSCluster( ... , services={('bokeh', ('', 8787)): BokehScheduler})
```

as a recommended solution -- on a machine like cheyenne, it's entirely possible that multiple users will be running jupyterlab on the same compute node, and only one user can claim `8787`. I think @guillaumeeb provides a superior option in [dask/dask-jobqueue#90](https://github.com/dask/dask-jobqueue/issues/90#issuecomment-410997419):
```
dask.config.set({'distributed.dashboard.link':'http://localhost:8888/proxy/{port}/status'})
cluster = PBSCluster( ... , ip='')
```

Also, to point out the obvious, this shouldn't be recommended until dask/dask-jobqueue#107 is available in a dask-jobqueue release. Currently setting `ip=''` via `PBSCluster` results in an error in all versions available via conda:

```
$ conda search dask-jobqueue
Fetching package metadata .............
dask-jobqueue                0.1.1                      py_0  conda-forge
                             0.2.0                      py_0  conda-forge
                             0.3.0                      py_0  conda-forge
```

**Comment by mnlevy1981** [comment]
> not sure if you noticed that on dask-jobqueue master branch, you don't even need extra args. Bokeh server is listenning on every network interfaces by default.

@guillaumeeb I didn't know that, thanks for pointing it out! I used the master branch to verify that nbserverproxy worked as advertised, but I think some other changes in the release required updating my `jobqueue.yaml` config file (IIRC what I believed was a request for five worker nodes submitted two single-node jobs to the queue), so I went back to using 0.3.0 and opening up a second ssh tunnel for the dashboard.

My plan is to wait for the next release before upgrading and then either [a] ask @jhamman or @mrocklin for the latest recommended configuration or [b] just see what changes they made in their own config files on the system - this hasn't been a huge priority for me.

**Comment by guillaumeeb** [comment]
> IIRC what I believed was a request for five worker nodes submitted two single-node jobs to the queue

Biggest change on master is https://github.com/dask/dask-jobqueue/pull/63. This allows adaptive cluster to work in most cases, but changes the arg of `scale` from `number of jobs` to `number of worker processes`.

**Comment by jhamman** [comment]
@mnlevy1981 - I know you've been looking into this issue a fair bit lately. I'm hoping you can bring this home when you have time.

**Comment by mnlevy1981** [comment]
@jhamman sorry to sit on this for a week, I was sidetracked by a few other things. If you can merge the latest master onto the branch, I can read through it and make a few suggestions.

**Comment by jhamman** [comment]
@mnlevy1981 - back in your court. Thanks for taking a look.

File: docs/setup_guides/hpc.rst (hunk 4 above)
Comment (mnlevy1981) [reply docs/setup_guides/hpc.rst]: Given the instructions in 216 ("Now open http://localhost:8888 on your local machine"), I think it's safe to recommend

```
diagnostics-link: "http://localhost:8888/proxy/{port}/status"
```

That's how we set it up in the [jobqueue.yaml on NCAR/pangeo](https://github.com/NCAR/pangeo/blob/master/server-side/jobqueue.yaml)

File: docs/setup_guides/hpc.rst [hunk 5]
```diff
@@ -222,12 +234,24 @@ done from within your Jupyter Notebook:
                          queue='premium',
                          resource_spec='select=1:ncpus=36:mem=109G',
                          walltime='02:00:00')
-    cluster.start_workers(10)
+    cluster.start_workers(18)
```
[the remark this answers is not in the document]
Comment (roxyboy) [reply docs/setup_guides/hpc.rst]: I don't know if it has to do with my version of dask-jobqueue but I get the following error with
```
cluster = PBSCluster(local_directory = '/glade/scratch/takayauc/ntbk-outputs/',
                             processes=18,
                             threads=4, memory="6GB",
                             project='UCLB0022',
                             queue='premium',
                             resource_spec='select=1:ncpus=36:mem=109G',
                             walltime='1:00:00')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-103654031b86> in <module>
      7                              queue='premium',
      8                              resource_spec='select=1:ncpus=36:mem=109G',
----> 9                              walltime='1:00:00')
     10 cluster.start_workers(16)
     11 

~/miniconda3/envs/uptodate/lib/python3.7/site-packages/dask_jobqueue/pbs.py in __init__(self, queue, project, resource_spec, walltime, job_extra, **kwargs)
     71 
     72         # Instantiate args and parameters from parent abstract class
---> 73         super(PBSCluster, self).__init__(**kwargs)
     74 
     75         # Try to find a project name from environment variable

~/miniconda3/envs/uptodate/lib/python3.7/site-packages/dask_jobqueue/core.py in __init__(self, name, cores, memory, processes, interface, death_timeout, local_directory, extra, env_extra, walltime, threads, **kwargs)
    161         # """
    162         if threads is not None:
--> 163             raise ValueError(threads_deprecation_message)
    164 
    165         if not self.scheduler_name:

ValueError: The threads keyword has been removed and the memory keyword has changed.

Please specify job size with the following keywords:

-  cores: total cores per job, across all processes
-  memory: total memory per job, across all processes
-  processes: number of processes to launch, splitting the quantities above
```

Using the following arguments work fine:
```
cluster = PBSCluster(local_directory = '/glade/scratch/takayauc/ntbk-outputs/',
                             cores=36,
                             processes=9,
                             memory="109GB",
                             project='UCLB0005',
                             queue='premium',
                             resource_spec='select=1:ncpus=36:mem=109G',
                             walltime='1:00:00')
```

**Pull request merged**